src/Http/Routing/src/Matching/DfaMatcherBuilder.cs

```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
                                 };
                             }
 
-                            // A parameter should traverse all literal nodes as well as the parameter node
                             if (parent.Literals != null)
                             {
-                                nextParents.AddRange(parent.Literals.Values);
+                                // If the parameter contains constraints, we can be smarter about it and evaluate them while we build the tree.
+                                // If the literal doesn't match any of the constraints, we can prune the branch.
+                                // For example, for a parameter in a route {lang:length(2)} and a parent literal "ABC", we can check that "ABC"
+                                // doesn't meet the parameter constraint (length(2)) when building the tree, and avoid the extra nodes.
+                                if (endpoint.RoutePattern.ParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicyReferences))
+                                {
+                                    // The list of parameters that fail to meet at least one ILiteralConstraint.
+                                    var hasFailingPolicy = new bool[parent.Literals.Keys.Count];
```

Comment:
stackalloc if `parent.Literals.Keys.Count` is smaller than 256? (which should be always, but allocate larger array just in case)

Reply:
Although, probably should only do that if this logic is in its own method. stackalloc in a loop doesn't behave well.

Reply:
this is now in its own method